blueprints/snapshots.and.monitoring.md

```diff
@@ -0,0 +1,828 @@
+Snapshot (and monitoring payload) data can be served in old API versions.
+
+### Client Impact
+The APIv4 client will need to change the call signatures of the affected
+endpoint methods.
+
+### Database Impact
+Existing stored Snapshots (and monitoring payload data) will need to be
+converted to the new format via a migration. No schema changes will be
+necessary, because Snapshots (and monitoring payload data) are stored as
+JSON blobs.
+
+## ORT Impact
+None.
+
+## Traffic Monitor Impact
+Traffic Monitor will need to be overhauled; Snapshot dependency is ingrained
+fairly deep in it. It will need to build out its configuration and stats from
+only the monitoring payload. It will also need to keep serving the Snapshots,
+however, to support old versions of Traffic Router.
+
+Traffic Monitor's parsing of monitoring payloads will also need to be updated
+to handle the new format - but as always, a previous API version must still be
+supported.
+
+## Traffic Router Impact
+Traffic Router will need to be updated to handle the new Snapshot format, as
+well as maintaining compatibility with the old format. It will also need to be
+updated to fetch Snapshots from Traffic Ops directly, rather than through
+Traffic Monitor.
+
+## Traffic Stats Impact
+None.
+
+## Traffic Vault Impact
+None.
+
+## Documentation Impact
+Documentation for CDN Snapshots and Monitoring payloads will need to be updated
+to reflect the new formats. This includes all of the affected endpoints'
+documentation, but should also include an overview section for each, now that
+the typing is more solidly defined.
+
+## Testing Impact
+Existing tests for each affected endpoint will need to be updated. Also, tests
+should be written for the logic for converting from the new formats to legacy
+formats, as well as for the new parsing logic in Traffic Monitor and Traffic
+Ops.
+
+## Performance Impact
+Many of the fields of Snapshots and Monitoring payloads have been changed from
+strings containing arbitrary values to statically typed properties. This
+generally means that Traffic Ops will need to spend some extra time parsing
+some properties - although sometimes the reverse is true where TO is currently
+transforming booleans into strings representing booleans. That will slightly
+increase the load on Traffic Ops.
+
+However, on the other side (Traffic Monitor and Traffic Router), less time will
+be spent doing the same parsing in reverse, so some performance gains will be
+made. Also, some mappings (most notably from a server's
+"cacheGroup"/"locationID" property to an entry in a 'cacheGroups' map) will be
+done on the Traffic Ops side, which incurs a negligible cost to TO itself but
+saves potentially significant time reproducing those mappings manually on the
+Traffic Monitor and Traffic Router side. Overall the expected performance
+impact is positive.
+
+## Security Impact
+There should be no changes from a security standpoint.
+
+## Upgrade Impact
```

Comment:
Will TM/TR first try to request the v4 API and failing that retry with v3 APIs calls?

Reply:
Yes. TM already does this (currently v3->v2). I don't know if TR already does but it should because all components are supposed to do that to maintain backward-compatibility.